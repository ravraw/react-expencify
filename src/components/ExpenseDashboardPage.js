import React from "react";
const ExpenseDashboardPage = () => (
  <div>
    <p>This is ExpenseDashboardPage</p>
  </div>
);

export default ExpenseDashboardPage;
