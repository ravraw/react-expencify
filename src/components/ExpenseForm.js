import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";

const now = moment();

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    console.log(props);
    this.state = {
      description: this.props.expense.description || "",
      note: this.props.expense.note || "",
      amount: (this.props.expense.amount / 100).toString() || "",
      createdAt: moment(this.props.expense.createdAt) || moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    this.setState({ description: e.target.value });
  };
  onNoteChange = e => {
    this.setState({ note: e.target.value });
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount });
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount"
      }));
    } else {
      this.setState(() => {
        error: "";
      });
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render(props) {
    console.log(this.props);
    return (
      <div>
        {this.state.error && <p style={{ color: "red" }}>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Expense"
            value={this.state.description}
            autoFocus
            onChange={e => this.onDescriptionChange(e)}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={e => this.onAmountChange(e)}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />

          <textarea
            name="note"
            id=""
            cols="30"
            rows="1"
            placeholder="Note for your expense (optional)"
            value={this.state.note}
            onChange={e => this.onNoteChange(e)}
          />
          <button>Submit</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
