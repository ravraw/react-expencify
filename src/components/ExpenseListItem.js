import React from "react";

import { Link } from "react-router-dom";

const ExpenseListItem = ({
  dispatch,
  id,
  description,
  note,
  amount,
  createdAt
}) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        {note} - {amount} - {createdAt}
      </p>
    </div>
  );
};

export default ExpenseListItem;
