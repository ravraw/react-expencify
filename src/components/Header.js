import React from "react";
import { NavLink } from "react-router-dom";
const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" activeClassName="active" exact>
      HOME
    </NavLink>
    <NavLink to="/add" activeClassName="active">
      ADD EXPENCE
    </NavLink>

    <NavLink to="/help" activeClassName="active">
      HELP
    </NavLink>
  </header>
);

export default Header;
