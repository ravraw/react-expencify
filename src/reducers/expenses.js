const expenseReducerDefaultState = [];
const expenseReducer = (state = expenseReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      // return state.concat(action.expense);
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(exp => {
        if (exp.id === action.id) {
          const updatedExp = {
            ...exp,
            ...action.updates
          };
          return updatedExp;
        }
        return exp;
      });
    default:
      return state;
  }
};

export default expenseReducer;
