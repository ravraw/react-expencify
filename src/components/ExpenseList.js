import React from "react";
import ExpenseListItem from "./ExpenseListItem";
import getFilteredExpenses from "../selectors/getFilteredExpenses";
import { connect } from "react-redux";

const ExpenseList = props => {
  const filteredExpenses = getFilteredExpenses(props.expenses, props.filters);

  const ExpList = filteredExpenses.map((el, index) => (
    <div key={el.id}>
      <ExpenseListItem {...el} />
    </div>
  ));

  return (
    <div>
      <h1>Expense LIST</h1>
      {ExpList}
      {props.filters.text}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: state.expenses,
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseList);
