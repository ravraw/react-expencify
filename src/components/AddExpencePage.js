import React from "react";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpencePage extends React.Component {
  onSubmit = expense => {
    this.props.onSubmit(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <p>Add expense page</p>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: expense => dispatch(addExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpencePage);
