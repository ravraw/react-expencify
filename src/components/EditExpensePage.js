import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

const EditExpensePage = props => {
  const expense = props.expenses.find(exp => exp.id === props.match.params.id);
  // console.log(expense);
  return (
    <div>
      <h3>{expense.description}</h3>
      <button
        onClick={() => {
          props.dispatch(removeExpense({ id: expense.id }));
          props.history.push("/");
        }}
      >
        Remove
      </button>

      <ExpenseForm
        expense={expense}
        onSubmit={updatedExp => {
          props.dispatch(editExpense(expense.id, updatedExp));
          props.history.push("/");
        }}
      />
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: state.expenses
  };
};

export default connect(mapStateToProps)(EditExpensePage);
