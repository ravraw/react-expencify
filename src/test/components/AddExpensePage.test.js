import React from "react";
import { AddExpencePage } from "../../components/AddExpencePage";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";

let onSubmit, history, wrapper;

beforeEach(() => {
  onSubmit = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <AddExpencePage onSubmit={onSubmit} history={history} id="form" />
  );
});

test("Should render correctly AddExpensePage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should handle onSubmit", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(onSubmit).toHaveBeenLastCalledWith(expenses[1]);
});
