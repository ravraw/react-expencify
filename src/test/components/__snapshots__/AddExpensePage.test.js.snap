// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly AddExpensePage 1`] = `
<div>
  <p>
    Add expense page
  </p>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
